service/s3/s3manager/batch.go

```diff
@@ -0,0 +1,225 @@
+package s3manager
+
+import (
+	"io"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+)
+
+// BatchDeleteIterator is an interface that uses the scanner pattern to
+// iterate through what needs to be deleted.
+type BatchDeleteIterator interface {
+	Next() bool
+	Err() error
+	DeleteObject() BatchDeleteObject
+}
+
+// BatchDelete will use the s3 package's service client to perform a batch
+// delete.
+type BatchDelete struct {
+	Client    s3iface.S3API
+	BatchSize int
+}
+
+// NewBatchDelete will return a new delete client that can delete a batched amount of
+// objects.
+//
+// Example:
+//	batcher := s3manager.NewBatchDelete(client, size)
+//
+//	objects := []BatchDeleteObject{
+//		{
+//			Object:	&s3.DeleteObjectInput {
+//				Key: aws.String("key"),
+//				Bucket: aws.String("bucket"),
+//			},
+//		},
+//	}
+//
+//	iter := s3manager.NewBatchDeleteIterator([]BatchDeleteObject{})
+//	if err := batcher.Delete(iter); err != nil {
+//		return err
+//	}
+func NewBatchDelete(client s3iface.S3API, size int) *BatchDelete {
+	return &BatchDelete{
+		Client:    client,
+		BatchSize: size,
+	}
+}
+
+func next(index, size int, err error) bool {
+	return index < size && err == nil
+}
+
+// BatchDeleteObject is a wrapper object for calling the batch delete operation.
+type BatchDeleteObject struct {
+	Object *s3.DeleteObjectInput
+}
+
+type batchDeleteIter struct {
+	Objects []BatchDeleteObject
+	index   int
+	err     error
+}
+
+func (batcher *batchDeleteIter) Next() bool {
+	return next(batcher.index, len(batcher.Objects), batcher.err)
+}
+
+func (batcher *batchDeleteIter) Err() error {
+	return batcher.err
+}
+
+func (batcher *batchDeleteIter) DeleteObject() BatchDeleteObject {
+	object := batcher.Objects[batcher.index]
+	batcher.index++
+	return object
+}
+
+// NewBatchDeleteIterator is the default iterator for the batch delete operation. This iterator
+// can only be used once.
+func NewBatchDeleteIterator(objects []BatchDeleteObject) BatchDeleteIterator {
+	return &batchDeleteIter{
+		Objects: objects,
+	}
+}
+
+// Delete will use the iterator to queue up objects that need to be deleted.
+// Once the batch size is met, this will call the deleteBatch function.
+func (d *BatchDelete) Delete(iter BatchDeleteIterator) error {
+	objects := make([]BatchDeleteObject, 0, d.BatchSize)
+	for iter.Next() {
+		objects = append(objects, iter.DeleteObject())
+
+		if len(objects) >= d.BatchSize {
+			if err := deleteBatch(d.Client, objects); err != nil {
+				return err
+			}
+			objects = objects[:0]
+		}
+	}
+
+	if err := iter.Err(); err != nil {
+		return err
+	}
+
+	return deleteBatch(d.Client, objects)
+}
+
+func deleteBatch(svc s3iface.S3API, objects []BatchDeleteObject) error {
```

Comment:
It doesn't look like this is really a batch download but an iterator based delete. Are these deletes concurrent? Id didn't see that if they are.